### PR TITLE
Adds Firestore sample

### PR DESCRIPTION
**Pull request by bshaffer**

### CODE REVIEW

File: firestore/src/firestore.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Google Inc.
```
Comment (wisdal) [inline firestore/src/firestore.kt]: Google LLC

File: firestore/src/firestore.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+package com.google.firestore
+
+import com.google.cloud.firestore.FirestoreOptions
+
+fun main(args: Array<String>) {
+    // validate the arguments
+    if (args.size < 1 || args.size > 3) {
```
Comment (wisdal) [inline firestore/src/firestore.kt]: args.isEmpty() rather than args.size < 1 ?

File: firestore/src/firestore.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        .setTimestampsInSnapshotsEnabled(true)
+        .build()
+        .service
+
+    // create the docRef and data object
+    val docRef = db.collection(args[0]).document("samples")
+    var data = docRef
+        .get() // future
+        .get() // snapshot
+        .getData() // MutableMap
```
Comment (wisdal) [inline firestore/src/firestore.kt]: .data (property access syntax)

File: firestore/src/firestore.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    // initialize document with empty data object if it doesn't exist
+    if (data == null) {
+        data = mutableMapOf<String, Any>()
+        docRef.set(data)
+    }
+
+    // If no arguments are supplied, call the quickstart. Fetch the key value if
+    // only one argument is supplied. Set the key to the supplied value if two
```
Comment (wisdal) [inline firestore/src/firestore.kt]: Can we refactor comments into 2 lines?

File: firestore/src/firestore.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    // arguments are supplied.
+    when (args.size) {
+        1 -> quickstart(args[0], "samples")
+        2 -> println("${args[1]}: " + (data.get(args[1]) ?: "not found"))
```
Comment (wisdal) [inline firestore/src/firestore.kt]: 2 -> println("${args[1]}: ${data[args[1]] ?: "not found"}")

File: firestore/src/quickstart.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
+// Copyright 2018 Google Inc.
```
Comment (wisdal) [inline firestore/src/quickstart.kt]: Google LLC.

File: firestore/src/quickstart.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+package com.google.firestore
+
+import com.google.cloud.firestore.FirestoreOptions
+
+fun quickstart(collectionName: String, documentName: String) {
+    // [START firestore_quickstart]
+    // import com.google.cloud.firestore.FirestoreOptions
```
Comment (wisdal) [inline firestore/src/quickstart.kt]: import comment to be deleted

File: firestore/src/quickstart.kt [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        .setTimestampsInSnapshotsEnabled(true)
+        .build()
+        .service
+
+    // Fetch the document reference and data object.
+    val docRef = db.collection(collectionName).document(documentName)
+    var data = docRef
+        .get() // future
+        .get() // snapshot
+        .getData() // MutableMap
```
Comment (wisdal) [inline firestore/src/quickstart.kt]: .data ?: error("Document $collectionName:$documentName not found") and delete next line

File: firestore/test/FirestoreTest.kt [hunk 9]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2018 Google Inc.
```
Comment (wisdal) [inline firestore/test/FirestoreTest.kt]: Google LLC

File: firestore/test/FirestoreTest.kt [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+import org.junit.After
+import org.junit.Assert
+import org.hamcrest.CoreMatchers.containsString
+import com.google.firestore.main
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+
+internal class FirestoreTest {
+
+    var outContent = ByteArrayOutputStream()
```
Comment (wisdal) [inline firestore/test/FirestoreTest.kt]: private?

File: firestore/test/FirestoreTest.kt [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    val originalOut = System.out
```
Comment (wisdal) [inline firestore/test/FirestoreTest.kt]: System.out!!

File: firestore/test/FirestoreTest.kt [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    val collection = System.getenv("FIRESTORE_COLLECTION")
```
Comment (wisdal) [inline firestore/test/FirestoreTest.kt]: I would do System.getenv("FIRESTORE_COLLECTION")!!

File: firestore/test/FirestoreTest.kt [hunk 13]
```diff
@@ -22,41 +22,41 @@ import java.io.PrintStream
...
 
     @Before
     fun setUpStreams() {
         System.setOut(PrintStream(outContent))
     }
 
     @Before
-    fun verifyCollection() {
-        if (collection.isNullOrEmpty()) {
+    fun `verify collection`() {
```
Comment (bshaffer) [inline firestore/test/FirestoreTest.kt]: holy crap that's cool...

**Comment by bshaffer** [comment]
@wisdal if this looks good to you, will you approve the PR and then I will merge it?

**Pull request merged**